rift/node.py

```diff
@@ -2469,87 +2517,116 @@ def spf_run_direction(self, spf_direction):
                 continue
             # Mark that we now have the best path to the destination.
             destination.best = True
+
             # If the destination is a node (i.e. its key is a system-id number rather than an IP
             # prefix), potentially add its neighbor nodes and its prefixes as a new candidate or
             # as a new ECMP path for an existing candidate.
             if isinstance(dest_key, int):
-                self.spf_add_candidates_from_node(dest_key, dest_cost, candidates, spf_direction)
-        # For south-bound SPF runs only, decide which prefixes need to be positively disaggregated
-        if spf_direction == constants.DIR_SOUTH:
+                self.spf_add_candidates_from_node(dest_key, dest_cost, candidates,
+                                                  spf_direction, special_for_neg_disagg)
+
+        # For south-bound SPF runs only (not for negative disaggregation special run),
+        # decide which prefixes need to be positively disaggregated
+        if spf_direction == constants.DIR_SOUTH and not special_for_neg_disagg:
             self.spf_mark_pos_disagg_prefixes()
-        # SPF run is done. Install the computed routes into the route table (RIB)
-        self.spf_install_routes_in_rib(spf_direction)
 
-    def spf_add_candidates_from_node(self, node_system_id, node_cost, candidates, spf_direction):
-        node_ties = self.node_ties(self.spf_use_tie_direction(node_system_id, spf_direction),
-                                   node_system_id)
-        if node_ties == []:
+        # SPF run is done.
+        if not special_for_neg_disagg:
+            # Install the computed routes into the route table (RIB) if run is normal
+            self.spf_install_routes_in_rib(spf_direction)
+        else:
+            # We don't want to install any route in the RIB if the SPF run is special for negative disaggregation
```

Comment:
"We don't want to install any route in the RIB if the SPF run is special"

That comment is a bit misleading. We may very well end up installing negative routes in the RIB, but the procedure for deciding which routes to add/remove to/from the RIB is very different.